src/dawn/Optimizer/Stencil.cpp

```diff
@@ -445,4 +445,23 @@ std::ostream& operator<<(std::ostream& os, const Stencil& stencil) {
   return os;
 }
 
+std::unordered_map<int, Extents> const& Stencil::computeEnclosingAccessExtents() {
```

Comment:
yes, since this computation is expensive, I guess computing into a member a returning by const& makes lot of sense. 
However the problem, is that this creates information that should belong to the internal IR, which is derived from other information that is generated from other Passes. And is difficult to know when this has to be recomputed and when is properly updated. Also there might be other derived states like the method `updateFields`

So I propose the following convention:
* ```void class::computeXXX``` just compute a derive state of the internal IR, in this case the map of accessID to extent. 
* ```auto class::getXXX``` is just a getter
* there is an update method that updates all the derived states of the IIR, i.e.
```
void update() {
  updateFields();
  computeEnclosingAccessExtents();
}
```

you need then to figure where you would need to apply this update, which is probably at the and of all the passes (unless one of the passes would require this information, which is not the case)

Reply:
i thought about the issue that this computation is expensive and should probably not computed everytime somebody is calling the method. i left it there because of the issue you mentioned (probably modified by a pass or so).

But nevertheless the code is only generated once and therefore the computeEnclosingAccessExtents is only called once (at the moment). Do you see any other case where the maximum enclosing extents are use beside the code gen?

Reply:
Maybe I didn't get it right. So please correct me if I'm wrong! The extent information is contained in the different stages. The first place were we can think or reason about "maximum extents" of several stages is a stencil. 

So I would personally vote for keeping this information in the stencil as I think it is more confusing and complicated to keep it outside. What do you mean with derived information?

**I've got another question regarding the update and expensiveness of the computation:** Is there any other case where the maximum enclosing extents are use beside the code gen? Is there any possibility that maximum enclosing extents are changing after the IR was generated?

Reply:
Yes, the first point where it makes sense to reason about that is a stencil. What happens though if we split a user-stencil [IStencil] (aka input) into two stencils being generated [GStencil]. There is a 1:n relationship between those two.
On what level is it useful to think about extents? I think it might be interetsing for the user to see if the field he passes to the driver is big enough for all the GStencils and therefore we should link it to IStencil [which is 1:1 with stencilInstantiaion]

I consider every result of an algorithm on data to be "derived" information. Currently we do our best to store those things in the StencilInstantiation

Not yet (otherwise the algorithm would have been in place :P) But I think it could happen in the future. @cosunae thoughts?
Depending on where we compute the maximum extents, it might happen. Think of what happens if this is the first thing we do and later we merge two stages (in the stage-merging pass), one of which has an extent, one does not. The one that did not was computed without and now has an extent.
There is where the automatic reordering / re-triggering of passes would be helpful. Assume we create a new pass that needs to be ran before stage-merging and requires full extent computation as input. we would need to run it before said pass and at the end once more.

https://github.com/MeteoSwiss-APN/dawn/blob/56af8b3eb43de26af4682f14fd3d4bcd5b4bf449/src/dawn/Compiler/DawnCompiler.cpp#L158
is included twice since this is not in place yet and has to be done manually